packages/block-library/src/gallery/index.php

```diff
@@ -0,0 +1,52 @@
+<?php
+/**
+ * Server-side rendering of the `core/image` block.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Handles backwards compatibility for Gallery Blocks,
+ * whose images feature a `data-id` attribute.
+ *
+ * Now that the Gallery Block contains inner Image Blocks,
+ * we add a custom `data-id` attribute before rendering the gallery
+ * so that the Image Block can pick it up in its render_callback.
+ *
+ * @param  array $parsed_block A single parsed block object.
+ *
+ * @return array               The migrated block object.
+ */
+function render_block_core_gallery_data( $parsed_block ) {
+	if ( 'core/gallery' === $parsed_block['blockName'] ) {
+		foreach ( $parsed_block['innerBlocks'] as $key => $inner_block ) {
+			if ( 'core/image' === $inner_block['blockName'] ) {
+				if ( ! $parsed_block['innerBlocks'][ $key ]['attrs']['data-id'] ) {
+					$parsed_block['innerBlocks'][ $key ]['attrs']['data-id'] = esc_attr( $inner_block['attrs']['id'] );
```

Comment:
Images that are linked from an external source don't have id attributes, so this can also throw a PHP notice:

![image](https://user-images.githubusercontent.com/14988353/140452928-e3d380b5-e009-4903-b20b-5269526d2e08.png)

Should we guard this one behind an `isset( $inner_block['attrs']['id'] )` check? 

To test, within a gallery block, select one of the images, and click Replace in the toolbar. Then, enter an external image, such as https://s.w.org/style/images/about/WordPress-logotype-standard.png as the image.

Reply:
Another good catch - done